resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java

```diff
@@ -100,11 +97,53 @@ protected DnsNameResolverContext(DnsNameResolver parent,
         allowedQueries = maxAllowedQueries;
     }
 
-    protected Promise<T> promise() {
-        return promise;
+    void resolve(Promise<T> promise) {
+        resolve(promise, parent.searchDomains().length > 0 && !hostname.endsWith("."));
     }
 
-    void resolve() {
+    private void resolve(Promise<T> promise, boolean trySearchDomain) {
+        if (trySearchDomain) {
+            final Promise<T> original = promise;
+            promise = parent.executor().newPromise();
+            promise.addListener(new FutureListener<T>() {
+                @Override
+                public void operationComplete(Future<T> future) throws Exception {
+                    if (future.isSuccess()) {
+                        original.setSuccess(future.getNow());
+                    } else {
+                        future.addListener(new FutureListener<T>() {
```

Comment:
the listener is reused to listen the various search domains `nextPromise.addListener(this)` until one returns a valid result or all have been tried